libclambcc/Common/ClamBCUtilities.cpp

```diff
@@ -146,3 +146,93 @@ llvm::BasicBlock *getEntryBlock(llvm::BasicBlock *pBlock)
 {
     return llvm::cast<llvm::BasicBlock>(pBlock->getParent()->begin());
 }
+
+int64_t getTypeSize(llvm::Module * pMod, llvm::Type *pt)
+{
+
+    int64_t size = pt->getScalarSizeInBits();
+    if (size) {
+        return size;
+    }
+
+    if (ArrayType *pat = llvm::dyn_cast<ArrayType>(pt)) {
+        size = pat->getNumElements() * (getTypeSize(pMod, pat->getElementType()));
+        if (size) {
+            return size;
+        }
+    }
+
+    if (StructType *pst = llvm::dyn_cast<StructType>(pt)) {
+        const StructLayout *psl = pMod->getDataLayout().getStructLayout(pst);
+        return psl->getSizeInBits();
+    }
+
+    assert(0 && "Size has not been computed");
+    return -1;
+}
+
+int64_t getTypeSizeInBytes(llvm::Module * pMod, Type *pt)
+{
+    return getTypeSize(pMod, pt) / 8;
+}
+
+int64_t computeOffsetInBytes(llvm::Module * pMod, Type *pt, uint64_t idx)
+{
+
+    int64_t cnt = 0;
+
+    assert((llvm::isa<StructType>(pt) || llvm::isa<ArrayType>(pt)) && "pt must be a complex type");
+
+    if (StructType *pst = llvm::dyn_cast<StructType>(pt)) {
+        assert((idx <= pst->getNumElements()) && "Idx too high");
+
+        const StructLayout *psl = pMod->getDataLayout().getStructLayout(pst);
+        assert(psl && "Could not get layout");
```

Comment:
Seeing a lot of asserts that will be compiled out in Release-mode builds or else cause crashes when they fail.  Do you feel confident that these will never fail and were only needed for debugging during development? 

Or could some bad bytecode C input result in these failures?  If so then maybe should be normal error handling instead of asserts.

Reply:
These asserts *should* all be coding errors in the compiler.  I will do more extensive testing using the llvm test suite before release, so I will re-evaluate them at that time.